rpc/src/rpc_subscriptions.rs

```diff
@@ -785,100 +842,168 @@ impl RpcSubscriptions {
-        let num_signatures_notified = AtomicUsize::new(0);
-
         let num_programs_found = AtomicUsize::new(0);
         let num_programs_notified = AtomicUsize::new(0);
 
+        let num_signatures_found = AtomicUsize::new(0);
+        let num_signatures_notified = AtomicUsize::new(0);
+
         let subscriptions = subscriptions.into_par_iter();
         subscriptions.for_each(|(_id, subscription)| {
+            let slot = if let Some(commitment) = subscription.commitment() {
+                if commitment.is_finalized() {
+                    Some(commitment_slots.highest_confirmed_root)
+                } else if commitment.is_confirmed() {
+                    Some(commitment_slots.highest_confirmed_slot)
+                } else {
+                    Some(commitment_slots.slot)
+                }
+            } else {
+                error!("missing commitment in notify_watchers");
+                None
+            };
             match subscription.params() {
                 SubscriptionParams::Account(params) => {
-                    let notified = check_commitment_and_notify(
-                        params,
-                        subscription,
-                        bank_forks,
-                        commitment_slots,
-                        |bank, params| bank.get_account_modified_slot(&params.pubkey),
-                        filter_account_result,
-                        notifier,
-                        false,
-                    );
-
                     num_accounts_found.fetch_add(1, Ordering::Relaxed);
-
-                    if notified {
-                        num_accounts_notified.fetch_add(1, Ordering::Relaxed);
+                    if let Some(slot) = slot {
+                        let notified = check_commitment_and_notify(
+                            params,
+                            subscription,
+                            bank_forks,
+                            slot,
+                            |bank, params| bank.get_account_modified_slot(&params.pubkey),
+                            filter_account_result,
+                            notifier,
+                            false,
+                        );
+
+                        if notified {
+                            num_accounts_notified.fetch_add(1, Ordering::Relaxed);
+                        }
+                    }
+                }
+                SubscriptionParams::Block(params) => {
+                    num_blocks_found.fetch_add(1, Ordering::Relaxed);
+                    if let Some(slot) = slot {
+                        let mut w_last_notified_slot =
+                            subscription.last_notified_slot.write().unwrap();
+                        // would mean it's the first notif
+                        if *w_last_notified_slot == 0 {
+                            *w_last_notified_slot = slot;
+                        }
+                        for s in *w_last_notified_slot..slot + 1 {
```

Comment:
I think this logic needs to be a little more complex. Some blocks between `[w_last_notified_slot, slot]` will be skipped by the cluster, and so should not be notified, even if they are complete in blockstore.
We need to get the list of ancestors of `slot` and only check the slots that are between `[w_last_notified_slot, slot]` and in the ancestor list.

Slot ancestors should be available via:
```
let ancestors = bank_forks.ancestors().get(slot);
```

Reply:
yea makes sense to me!